binfmtc-interpreter.1
.TH "binfmtc" 1 "2009 Feb 8" "binfmt_misc Dancer" "binfmt_C"
.SH NAME
binfmtc-interpreter \- binfmt_misc C handler
.SH SYNOPSIS
.BI "binfmtc-interpreter " "C-source-file-name" " [" "command-line opions ..." "]"
.SH "DESCRIPTION"
.B "binfmtc-interpreter"
compiles a C source file specified on the command-line using 
C compiler, and executes the resulting file.

It is designed to be used as a handler for 
.B "binfmt_misc"
handler, which is a system used in Linux for handling arbitrary files 
as executables.

The command-line options are passed on to the 
compiled binary.

.SH "FILE MAGIC"

There is a requirement for C source files to have the 
magic characters
.B "/*BINFMTC:"
at the beginning of the file.

That line also is used to specify the additional command-line options
for C compiler.

.SH "ENVIRONMENT"
.TP
.B "GCC"
The compiler used.
The default is to use
.B "gcc"

.TP
.B "CC"
The compiler used, if 
.B "GCC"
variable is not set.

.TP
.B "BINFMTCTMPDIR"

Temporary directory used for binary and execution.

Falls back to 
.B "$TMPDIR" 
.B "$TEMPDIR"
or
.B "/tmp"

.TP
.B "BINFMTC_DEBUG"
enables debug output if set.

.TP
.B "BINFMTC_GCC_OPTS"
Additional GCC options.
Use 
.B "BINFMTC_DEBUG"
to verify the options being passed on to gcc.

The default is
.B " -O2 -Wall -g "

.SH "AUTHOR"
Junichi Uekawa (dancer@debian.org)

Upstream page is available at 
.B "http://www.netfort.gr.jp/~dancer/software/"

.SH "SEE ALSO"
.BR "binfmtasm-interpreter" "(1), "
.BR "binfmtcxx-interpreter" "(1), "
.BR "binfmtf-interpreter" "(1), "
.BR "binfmtf95-interpreter" "(1), "
.BR "binfmtgcj-interpreter" "(1)" 



NEWS
0.17 21 Sep 2011

+ -g option added to default for C++/C for easy debugging.
+ realksh.c kernel printk message handling is more deterministic.
! removed pascal support.

0.16 4 Oct 2008

+ g77 to f95 (gfortran) transition, g77 is no longer available in
  Debian.
! Handle modules.order in realksh.c

0.15 2 Sep 2007

+ Remove temporary files at process exit time. 
  Process exit is monitored every second
! Handle Modules.symvers->Module.symvers change in Linux Kernel.

0.14 10 May 2007

! Fix typo in realcxxsh.cc.1 manpage
+ realcxxsh.cc improvement.

0.13 13 Jan 2007

+ realcxxsh.cc
+ realcsh.c supports adding gcc command-line options, so additional
  libs can be used in the interactive command-line

0.12 3 Sep 2006

! Memleak fix in realcsh.

0.11 5 Aug 2006

! Minor update on realksh cleanup routine, so that it cleans up
properly.

0.10 6 Jan 2006

+ add realksh.c sample program for kernel shell.
+ add a testsuite for realksh. 
  It's sometimes risky to run this on production kernels, beware.

0.9 9 Nov 2005

+ add testsuite for amd64 assembly

0.8 30 Aug 2005

! Add clean-up and error checking.

0.7 29 Jul 2005

+ Update binfmtc-init.sh and binfmtc-unreg.sh
+ realcsh.c supports ## and #include commands.

0.6 19 June 2005

+ add testsuite for x86 assembly

0.5 4 June 2005

+ Add Fortran support
+ Add Pascal support

0.4 4 June 2005

! Use daemon properly so that we don't have stale process.

0.3 24 May 2005

+ Add Java support

0.2 23 May 2005

+ Debian package

0.1 7 May 2005

+ Create Initial distribution

[KEYS
+: additional feature
!: bugfixes
-: obsoletion
]


binfmtc-interpreter.c
/*
 *  binfmt_misc C Interpreter
 *  Copyright (C) 2005 Junichi Uekawa
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * Interepreter for binfmt_misc C compilation.
 *
 */
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <ctype.h>
#include <sys/wait.h>
#include "binfmtc.h"

#define BINFMTC_MAGIC_LEN strlen(binfmtc_magic)
#define BINFMTC_DEBUG (getenv("BINFMTC_DEBUG"))

char* compile_source(const char *sourcename)
{
  char* origdir = getcwd(NULL, 0);
  char* path = strdup (sourcename);
  char* path_delimiter;
  char* tempfilename = NULL; 
  char* s = NULL;
  char* gcccommandline = NULL;
  size_t size;
  FILE* f;
  int i;
  
  asprintf(&tempfilename, "%s/binfmtcXXXXXX",
	   getenv("BINFMTCTMPDIR")?:
	   getenv("TMPDIR")?:
	   getenv("TEMPDIR")?:
	   "/tmp"
	   );
  
  if (NULL==(path_delimiter=strrchr(path, '/')))
    {
      fprintf(stderr, 
	      "binfmtc: Could not determine the directory name of source %s\n", 
	      sourcename);
      return NULL;
    }
  
  *path_delimiter=0;/* obtain the dirname */
  close(mkstemp(tempfilename)); /* get temporary filename */
  if (!(f=fopen(sourcename,"rt")))
    {
      fprintf(stderr, 
	      "binfmtc: Could not open file %s for read\n", 
	      sourcename);
      unlink(tempfilename);
      return NULL;
    }

  if (!getline (&s, &size, f))
    {
      fprintf(stderr, 
	      "binfmtc: Could not read file %s\n", 
	      sourcename);
      unlink(tempfilename);
      return NULL;
    }

  /* compare with magic */
  if (size < BINFMTC_MAGIC_LEN || 
      (memcmp(binfmtc_magic, s, BINFMTC_MAGIC_LEN)))
    {
      fprintf(stderr, 
	      "binfmtc: %s magic invalid \n", 
	      sourcename);
      unlink(tempfilename);
      return NULL;
    }
  fclose(f);

  /* make the newlines removed */
  for(i=strlen(s)-1; i; --i)
    {
      if (!isspace (s[i]))
        break;
      else
	s[i]=0;
    }

  asprintf (&gcccommandline,
	    "%s -o %s %s %s %s \"%s\"",
	    compiler_name(),
	    tempfilename,
	    gcc_x,
	    s+BINFMTC_MAGIC_LEN,
	    default_options(),
	    basename(sourcename));

  if (BINFMTC_DEBUG)
    fprintf(stderr, "binfmtc: Execute command-line: %s\n", gcccommandline);

  if (chdir(path))			/* go to the source's dir */
    {
      fprintf(stderr, 
	      "binfmtc: Cannot chdir to directory where source is: %s \n", 
	      path);
      unlink(tempfilename);
      return NULL;
    }
  
  if (system(gcccommandline))
    {
      fprintf(stderr, 
	      "binfmtc: Compilation failed for %s, see above messages for details\n", 
	      sourcename);
      unlink(tempfilename);
      return NULL;
    }
  
  if (chdir(origdir))
    {
      fprintf(stderr, 
	      "binfmtc: Cannot chdir to original working directory: %s \n", 
	      path);
      unlink(tempfilename);
      return NULL;
    }

  free(origdir);
  free(path);
  free(s);
  free(gcccommandline);
  return tempfilename;
}

/**
   Function to execute the compiled program.
   
   The reason the compiled program is execed as the parent program
   is to allow use of same PID for the compiled program, so that 
   it is possible to use "strace" "kill -HUP" and other conventional 
   methods of debugging.
 */
int exec_prog(const char * filename, int argc, char**argv)
{
  int pid;
  int parent_pid;

  switch(pid=fork())
    {
    case -1:
      /* fork failed */
      fprintf(stderr, 
	      "binfmtc: Failed to fork \n");
      return EXIT_FAILURE;
      break;
    case 0:
      /* when I am the child process, unlink the file after the parent
	 process has exited.
       */
      parent_pid=getppid();
      if (daemon(0,0) < 0)
	{
	  exit (-1);
	}
      while ((parent_pid!=-1)
	     && (kill(parent_pid,0) >= 0))
	{
	  sleep(1);
	}
      unlink(filename);
      exit(0);
    default:
      /* I am the parent process, exec into the built program */
      waitpid(pid,NULL,0);
      execvp(filename, argv);
      fprintf(stderr, 
	      "binfmtc: failed execvp of %s \n", filename);
      break;
    }
  return EXIT_FAILURE;
}

int main(int argc, char**argv)
{
  char* filename=NULL;
  int exit_code = EXIT_FAILURE;
  if (argc < 2)
    {
      /* Wrong number of command-line args; maybe someone tried to exec me directly. */
      fprintf(stderr,
	      "binfmtc: Wrong number of command-line options, use binfmt_misc\n%s\n",
	      language_type);
    }
  else if ((filename=compile_source(argv[1])))
    exit_code=exec_prog(filename, argc-1, &(argv[1]));
  return exit_code;
}
